certbot-nginx/certbot_nginx/configurator.py

```diff
@@ -565,6 +566,45 @@ def _has_certbot_redirect(self, vhost, domain):
         test_redirect_block = _test_block_from_block(_redirect_block_for_domain(domain))
         return vhost.contains_list(test_redirect_block)
 
+    def _should_break_block(self, vhost):
+        # if host has both SSL and non-SSL addrs, we should break this block!
+        return vhost.ssl and any([not addr.ssl for addr in vhost.addrs])
+
+    def _set_http_header(self, domain, header_substring):
+        """Enables header identified by header_substring on domain.
+
+        If the vhost is listening plaintextishly, separates out the relevant
+        directives into a new server block, and only add header directive to
+        HTTPS block.
+
+        :param str domain: the domain to enable header for.
+        :param str header_substring: String to uniquely identify a header.
+                        e.g. Strict-Transport-Security, Upgrade-Insecure-Requests
+        :returns: Success
+        :raises .errors.PluginError: If no viable HTTPS host can be created or
+            set with header header_substring.
+        """
+        vhost = self.choose_vhost(domain)
+        if vhost.has_header(header_substring):
+            raise errors.PluginEnhancementAlreadyPresent(
+                "Existing %s header" % (header_substring))
+
+        if vhost is None:
+            # this should never happen, since enhancements occur after
+            # setting certs!
+            raise errors.PluginError(
+                "Unable to find corresponding HTTP host for enhancement.")
+
+        # if there is no separate SSL vhost, break the vhost into two blocks and
+        # choose the SSL vhost.
+        if self._should_break_block(vhost):
+            _, vhost = self._split_block(vhost)
```

Comment:
I know I just said we can actually reuse code, but we maybe actually can't quite as much. `_split_block` for redirect keeps only the `server_name` and `listen` directives in the new http host; here, we probably want to copy over all the directives, keeping `default_server` if it's there (it should do that automatically but I'd double-check), and then just removing the appropriate `listen` directives. It shouldn't be more of a change than passing in an optional `only_directives` parameter from the outside.

Reply:
does it affect the HTTP block if it's got other ssl configuration directives set but not `listen ssl`?

Reply:
the fun thing about nginx is that there's not so much on the documentation and you usually just have to try it to see if it works; do you have a test config setup on a box somewhere?

Reply:
just tested -- seems to work fine. it might be more aesthetically pleasing to remove the ssl directives, what do you think?

Reply:
agreed-- it makes more sense not to include ssl directives. just did this-- let me know if i missed anything!